[Jason Qiu] iP

jasonqiu212 opened the pull request:
# Cheese

> Dogs are not our _whole_ life, but they make our lives _whole_. 

Cheese is a 🐶-themed task manager! It's...
- Command-line based (for the tech-savvy ones out there)
- **Simple** to use
- ~~Stressful~~ Stress-free to use with puppy phrases

### Features

- [x] Manage tasks
- [x] Manage deadlines and events
- [ ] Graphical user interface

## Set-up Guide
1. Download the latest [release](https://github.com/jasonqiu212/ip/releases/tag/A-Jar).
2. Click on it.
3. Add your tasks right away!

## Developer's Guide

For budding Java programmers, feel free to practice your skills on this repository! The entry point is at `Cheese.java`.
```java
public static void main(String[] args) {
    new Cheese("data/cheese.txt").run();
}
```

Commit:
Add Gradle support

jasonqiu212 committed:
feat: add greeting

jasonqiu212 committed:
feat: add respond feature

jasonqiu212 committed:
refactor: refactor chat function into separate method

jasonqiu212 committed:
feat: add list feature

jasonqiu212 committed:
feat: add Task class to store task description and status

jasonqiu212 committed:
feat: add feature to mark task as done/undone

jasonqiu212 committed:
feat: add Event and Deadline and Todo class

jasonqiu212 committed:
refactor: rename file class name to Duke

jasonqiu212 committed:
feat: add feature of adding Todo and Deadline and Event

jasonqiu212 committed:
refactor: rename file class name to Cheese

jasonqiu212 committed:
feat: edit input and expected file for testing

jasonqiu212 committed:
refactor: edit chatbot messages

jasonqiu212 committed:
refactor: use switch statements to determine command

jasonqiu212 committed:
feat: add error handling

jasonqiu212 committed:
refactor: refactor error handling

jasonqiu212 committed:
fix: fix create event

jasonqiu212 committed:
feat: add new tests to test error handling

jasonqiu212 committed:
fix: fix .gitignore typo

jasonqiu212 committed:
feat: add delete feature

jasonqiu212 committed:
feat: add test for delete feature

jasonqiu212 committed:
refactor: refactor commands using enums

jasonqiu212 committed:
docs: add documentation

jasonqiu212 committed:
refactor: add protected constructor for Task

jasonqiu212 committed:
Add feature to load tasks from .txt file

jasonqiu212 committed:
Add feature to save tasks to hard disk

jasonqiu212 committed:
Add support for datetime

jasonqiu212 committed:
Merge branch 'branch-Level-7'

jasonqiu212 committed:
Merge branch 'branch-Level-8'

jasonqiu212 committed:
Add storage class

jasonqiu212 committed:
Refactor Cheese and TaskList with Ui

jasonqiu212 committed:
Refactor Task with Ui

jasonqiu212 committed:
Add ByeCommand

jasonqiu212 committed:
Refactor ListCommand

jasonqiu212 committed:
Create MarkCommand and UnmarkCommand and DeleteCommand

jasonqiu212 committed:
Create TodoCommand and DeadlineCommand and EventCommand

jasonqiu212 committed:
Create UnknownCommand to handle unknown commands

jasonqiu212 committed:
Edit dates in Cheese data file

jasonqiu212 committed:
Fix readCommand bug

jasonqiu212 committed:
Add showGoodbye message function to ByeCommand

jasonqiu212 committed:
Fix index of getTask

jasonqiu212 committed:
Refactor dateTime methods to DateTimeUtils

jasonqiu212 committed:
Refactor into packages

jasonqiu212 committed:
Merge branch 'add-gradle-support'

jasonqiu212 committed:
Refactor package name into build.gradle

jasonqiu212 committed:
Add test for TaskList

jasonqiu212 committed:
Add test for Todo

jasonqiu212 committed:
Edit jar file name in ShadowJar

jasonqiu212 committed:
Add JavaDoc for commands

jasonqiu212 committed:
Add JavaDoc for data and exception and parser and storage and task

jasonqiu212 committed:
Add JavaDoc for ui and utils and Cheese.java

jasonqiu212 committed:
Fix JavaDoc for Parser and Storage

jasonqiu212 committed:
Fix indentation

jasonqiu212 committed:
Fix ordering of import statements

jasonqiu212 committed:
Merge branch 'branch-A-JavaDoc'

jasonqiu212 committed:
Merge branch 'branch-A-CodingStandard'

jasonqiu212 committed:
Find command: add find feature

There's no way for users to find task.

Let's
* add command to support searching tasks
* add methods in Task and TaskList to support searching

jasonqiu212 committed:
Merge branch 'branch-Level-9'

jasonqiu212 committed:
Find command: change search to case insensitive

Find is case sensitive.

Users may not remember the exact capitalization of their tasks. Thus, a case
insensitive find is more user-friendly.

Let's
* update search to use case insensitive matching

ZhengLong3 commented on src/main/java/cheese/task/Task.java:
```diff
@@ -0,0 +1,72 @@
+     * @param keyword Keyword to search for.
+     * @return True, if description contains given keyword. False otherwise.
+     */
+    public boolean hasKeyword(String keyword) {
```
I like that your boolean method sounds like a boolean.

jasonqiu212 replied:
thank you ZhengLong for reviewing my code!

ZhengLong3 commented on src/main/java/cheese/storage/Storage.java:
```diff
@@ -0,0 +1,95 @@
+     * @return Task parsed from string.
+     * @throws CheeseException If string from save file is corrupted.
+     */
+    private Task decodeStringToTask(String taskString) throws CheeseException {
```
I like how your methods are very descriptive in what they do. This can be seen in many other instances.

ZhengLong3 reviewed (commented):
The code looks very good to me.
All the methods are documented and the names are all very well chosen.
I also cannot find any coding standard violations.

gunbux commented on src/main/java/cheese/command/ByeCommand.java:
```diff
@@ -0,0 +1,30 @@
+     */
+    @Override
+    public void execute(TaskList taskList, Storage storage, Ui ui) {
+        ui.showGoodbye();
```
Something that could be done, is that since it seems that you are only using `ui` to print certain strings, maybe you could use `ui` as a static class, that way you don't need to keep parsing it in as a variable, you could just do `Ui.showGoodbye()`

jasonqiu212 replied:
that's a good idea! I'll refactor this for the Gui

gunbux commented on src/main/java/cheese/Cheese.java:
```diff
@@ -0,0 +1,66 @@
+        try {
+            taskList = storage.load();
+        } catch (CheeseException e) {
+            ui.showError(e.getMessage());
```
Although we're only supposed to comment on styling issues, maybe an alternative I'd like to suggest that is seems like you could probably use `Ui` as a static class, that way you don't need to keep parsing it into different parts of the code as an argument!

gunbux reviewed (commented):
Looks really great! Everything was very well named and organized, it was basically self documenting code!

gunbux commented on src/main/java/cheese/data/TaskList.java:
```diff
@@ -0,0 +1,151 @@
+     * @param taskIndex Index of task to validate.
+     * @throws CheeseException If index is not in the range of task list.
+     */
+    private void validateTaskIndexInRange(int taskIndex) throws CheeseException {
+        if (taskIndex < 0 || taskIndex >= taskList.size()) {
+            throw new CheeseException("Item number is not in list range.");
+        }
```
Really like how consistent your naming conventions are (i.e. `validate` for functions that throw errors), makes the code really easy to understand!

jasonqiu212 replied:
thanks for reviewing my code!

jasonqiu212 committed:
Add Java FX dependency to Gradle

jasonqiu212 committed:
Add GUI

jasonqiu212 committed:
Change mainClassName for gradle

jasonqiu212 committed:
Add fxml files for DialogBox and MainWindow

jasonqiu212 committed:
Add controllers for DialogBox and MainWindow

jasonqiu212 committed:
Remove text-based user interface

jasonqiu212 committed:
Add Response class to return messages from Cheese

jasonqiu212 committed:
Refactor commands to return response message

jasonqiu212 committed:
Move Ui files into package

jasonqiu212 committed:
Edit javadocs for gui files

jasonqiu212 committed:
Merge branch 'branch-Level-10'

jasonqiu212 committed:
Fix style

jasonqiu212 committed:
Refactor magic values to named constants

jasonqiu212 committed:
Refactor long expressions into boolean values

jasonqiu212 committed:
Storage: create directory for save file if it does not exist

There is no check for if the directory containing save file exists. This leads to
potential bugs if the directory is not present and we attempt to access the save file.

Users are not expected nor instructued to create this directory. Thus, we need to create
the directory if it does not exist.

Let's
* add check if directory exists
* create directory if it does not exist

jasonqiu212 committed:
Storage: create directory or file if nonexistent when saving

There is no check if the save file or directory exists when saving. There
is such a check for loading a file.

Successfully loading the save file at the start does not ensure the file
will remain when saving to it later. Thus, we need to perform the same
check when saving.

Let's,
* abstract out the logic of creating the new save file if nonexistent
* call the validation method before saving and loading

jasonqiu212 committed:
Storage: add assertion to ensure save file exists

jasonqiu212 committed:
Merge pull request #2 from jasonqiu212/branch-A-Assertions

Add assertions

jasonqiu212 committed:
Merge branch 'master' into branch-A-CodeQuality

jasonqiu212 committed:
Merge pull request #3 from jasonqiu212/branch-A-CodeQuality

Fix code quality

jasonqiu212 committed:
Fix javadocs

jasonqiu212 committed:
Add snooze command to postpone events and deadlines by 1 day

jasonqiu212 committed:
Merge pull request #4 from jasonqiu212/branch-B-Snooze

Add snooze command

jasonqiu212 committed:
Add CI/CD

jasonqiu212 committed:
Merge pull request #5 from jasonqiu212/branch-A-CI

Add CI/CD

jasonqiu212 committed:
Remove IO redirection style testing

jasonqiu212 committed:
Add style to dialog box and image

jasonqiu212 committed:
Add placeholder to text field

jasonqiu212 committed:
Edit style of dialog box

jasonqiu212 committed:
Merge pull request #6 from jasonqiu212/branch-A-BetterGui

Update Gui

jasonqiu212 committed:
Change images for Cheese and user avatar

jasonqiu212 committed:
Edit responses from Cheese

jasonqiu212 committed:
Increase size of dialog box image

jasonqiu212 committed:
Add title text to window

jasonqiu212 committed:
Merge pull request #7 from jasonqiu212/branch-A-Personality

Add personality

jasonqiu212 committed:
Add representative screenshot

jasonqiu212 committed:
Add general format for user guide

jasonqiu212 committed:
Add remaining commands to user guide

jasonqiu212 committed:
Add links to commands

jasonqiu212 committed:
Add features to user guide

jasonqiu212 committed:
Merge branch 'master' into branch-A-UserGuide

jasonqiu212 committed:
Merge pull request #8 from jasonqiu212/branch-A-UserGuide

Add user guide

jasonqiu212 committed:
Add checkstyle xml files

jasonqiu212 committed:
Add gradle support for checkstyle

jasonqiu212 committed:
Fix checkstyle errors

jasonqiu212 committed:
Merge pull request #9 from jasonqiu212/branch-A-Checkstyle

Add checkstyle

jasonqiu212 committed:
Edit docs

jasonqiu212 committed:
Update docs emojis

jasonqiu212 committed:
Update README.md

jasonqiu212 committed:
Update Ui.png

jasonqiu212 committed:
Update Ui.png
